server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java

```diff
@@ -380,16 +380,21 @@ public boolean syncUpTo(long offset) throws IOException {
 
     @Override
     protected void readBytes(ByteBuffer targetBuffer, long position) throws IOException {
-        if (position + targetBuffer.remaining() > getWrittenOffset()) {
-            synchronized (this) {
-                // we only flush here if it's really really needed - try to minimize the impact of the read operation
-                // in some cases ie. a tragic event we might still be able to read the relevant value
-                // which is not really important in production but some test can make most strict assumptions
-                // if we don't fail in this call unless absolutely necessary.
-                if (position + targetBuffer.remaining() > getWrittenOffset()) {
-                    outputStream.flush();
+        try {
+            if (position + targetBuffer.remaining() > getWrittenOffset()) {
+                synchronized (this) {
+                    // we only flush here if it's really really needed - try to minimize the impact of the read operation
+                    // in some cases ie. a tragic event we might still be able to read the relevant value
+                    // which is not really important in production but some test can make most strict assumptions
+                    // if we don't fail in this call unless absolutely necessary.
+                    if (position + targetBuffer.remaining() > getWrittenOffset()) {
+                        outputStream.flush();
+                    }
                 }
             }
+        } catch (final IOException e) {
+            closeWithTragicEvent(e);
+            throw e;
```

Comment:
Here, I think you should adopt the pattern we use throughout the rest of the class, i.e.,

```
try {
    closeWithTragicEvent(e);
} catch (Exception inner) {
    ex.addSuppressed(inner);
}
throw e;
```

so that we get the original cause.

Reply:
Bah! Thrown off by the fact that the method this is in declares `throws IOException`. I will address.

Reply:
I pushed 8ccaaf374fe4d10b95bba7f501238b5aabbdef27.